src/audio/mfcc/mfcc.c

```diff
@@ -0,0 +1,373 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <sof/audio/mfcc/mfcc_comp.h>
+#include <sof/audio/component.h>
+#include <sof/audio/data_blob.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/ipc-config.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/ipc/msg.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/mfcc.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_mfcc;
+
+LOG_MODULE_REGISTER(mfcc, CONFIG_SOF_LOG_LEVEL);
+
+/* db10a773-1aa4-4cea-a21f-2d57a5c982eb */
+DECLARE_SOF_RT_UUID("mfcc", mfcc_uuid, 0xdb10a773, 0x1aa4, 0x4cea,
+		    0xa2, 0x1f, 0x2d, 0x57, 0xa5, 0xc9, 0x82, 0xeb);
+
+DECLARE_TR_CTX(mfcc_tr, SOF_UUID(mfcc_uuid), LOG_LEVEL_INFO);
+
+const struct mfcc_func_map mfcc_fm[] = {
+#if CONFIG_FORMAT_S16LE
+	{SOF_IPC_FRAME_S16_LE,  mfcc_s16_default},
+#endif /* CONFIG_FORMAT_S16LE */
+#if CONFIG_FORMAT_S24LE
+	{SOF_IPC_FRAME_S24_4LE, NULL},
+#endif /* CONFIG_FORMAT_S24LE */
+#if CONFIG_FORMAT_S32LE
+	{SOF_IPC_FRAME_S32_LE,  NULL},
+#endif /* CONFIG_FORMAT_S32LE */
+};
+
+static mfcc_func mfcc_find_func(enum sof_ipc_frame source_format,
+				enum sof_ipc_frame sink_format,
+				const struct mfcc_func_map *map,
+				int n)
+{
+	int i;
+
+	/* Find suitable processing function from map. */
+	for (i = 0; i < n; i++) {
+		if (source_format != map[i].source)
+			continue;
+
+		return map[i].func;
+	}
+
+	return NULL;
+}
+
+/*
+ * End of MFCC setup code. Next the standard component methods.
+ */
+
+static struct comp_dev *mfcc_new(const struct comp_driver *drv,
+				 struct comp_ipc_config *config,
+				 void *spec)
+{
+	struct ipc_config_process *ipc_mfcc = spec;
+	struct comp_dev *dev = NULL;
+	struct mfcc_comp_data *cd = NULL;
+	size_t bs = ipc_mfcc->size;
+	int ret;
+
+	comp_cl_info(&comp_mfcc, "mfcc_new()");
+
+	/* Check first that configuration blob size is sane */
+	if (bs > SOF_MFCC_CONFIG_MAX_SIZE) {
+		comp_cl_err(&comp_mfcc, "mfcc_new() error: configuration blob size %u exceeds %d",
+			    bs, SOF_MFCC_CONFIG_MAX_SIZE);
+		return NULL;
+	}
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
```

Comment:
shall we use same exit flow, both goto fail or cd_fail?

Reply:
I wonder which one is less code size, a goto or return NULL here? We should use practices with smaller code size.

Reply:
code size impact should be minor, the intention is to keep same exit flow(port) in one function, since later code are all use go to for exit